odoo/models.py

```diff
@@ -6347,34 +6347,55 @@ def process(model, id_vals):
 
         for model_name, fields_ in model_fields.items():
             dirty_fields = self.env.cache.get_dirty_fields()
-            if any(field in dirty_fields for field in fields_):
-                # if any field is context-dependent, the values to flush should
-                # be found with a context where the context keys are all None
-                context_none = dict.fromkeys(
-                    key
-                    for field in fields_
-                    for key in self.pool.field_depends_context[field]
-                )
-                model = self.env(context=context_none)[model_name]
-                id_vals = defaultdict(dict)
-                for field in model._fields.values():
-                    ids = self.env.cache.clear_dirty_field(field)
-                    if not ids:
+            if not any(field in dirty_fields for field in fields_):
+                continue
+
+            # if any field is context-dependent, the values to flush should
+            # be found with a context where the context keys are all None
+            context_none = dict.fromkeys(
+                key
+                for field in fields_
+                for key in self.pool.field_depends_context[field]
+            )
+            model = self.env(context=context_none)[model_name]
+
+            # pop dirty fields and their corresponding record ids from cache
+            dirty_field_ids = {
+                field: set(ids)
+                for field in model._fields.values()
+                if (ids := self.env.cache.clear_dirty_field(field))
+            }
+            # get a reference to each dirty field's cache (containing the values)
+            dirty_field_cache = {
+                field: self.env.cache._get_field_cache(model, field)
+                for field in dirty_field_ids
+            }
+
+            # build a mapping {vals: ids} of field updates and their record ids
+            seen_ids = []
```

Comment:
Shouldn't seen_ids be a `set`? with `seen_ids.update(dirty_field_ids[field])`

Reply:
Even though it looks better, the memory use get's worse, because this will store another reference on each id. By only remembering the processed fields, `seen_ids` stays small (few fields, many ids).

Reply:
That merits a comment IMO. I guess the individual `dirty_field_ids[field]` are not that many so this "list of sets" is still well performant vs "only one set".